### ISSUE TITLE
bug: Incorrect parent and firstChild set for leftmost and rightmost nonterminals of zero length

### ISSUE THREAD

**Issue by mewmew** [inspirer/textmapper#23]
Creating a dedicated issue to track the bug identified at https://github.com/inspirer/textmapper/pull/17#issuecomment-433245585

I'll add my latest debug session below, feel free to skim as it's quite long :)

The debug session below is debugging the [none](https://github.com/mewspring/foo/tree/master/none/cmd/none) command, which uses the [mini.tm](https://github.com/mewspring/foo/blob/master/none/mini/mini.tm) grammar, and is invoked on the [example.ll](https://github.com/mewspring/foo/blob/master/none/example.ll) input file.

Specifically the following commands may be used to reproduce the debug session:

```bash
$ go get -u github.com/mewspring/foo/none/cmd/none
$ go get -u github.com/derekparker/delve/cmd/dlv
$ go get -u github.com/aarzilli/gdlv/
$ cd $GOPATH/src/github.com/mewspring/foo/none/cmd/none
$ $GOPATH/bin/gdlv debug ../../example.ll
```

Cheers,
Robin

### Debug session

Is it correct that next should be overwritten here?
Then we will skip `/* empty */` nonterminals that are in between other nonterminals.

![2018-10-27-010734_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597166-add70980-d98c-11e8-82e6-5ac1cd9d9fe0.png)

Should it really be `o >= endoffset`? And not `o > endoffset`? Now, we decrease the end, so that InstMetadata will not be part of the call instruction.

![2018-10-27-011328_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597173-b16a9080-d98c-11e8-920f-3f7b2e27b5ad.png)

OperandBundles and InstMetadata were not assigned to have CallInst as parent. They should have been. Note, end was too small. Most likely as a cause of using `if o >= offset { end-- }` instead of `if o > offset { end-- }`

![2018-10-27-011841_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597175-b4658100-d98c-11e8-84a8-4b28cad8fde8.png)

This is probably wrong, as now OperandBundles and InstMetadata won't have CallInst as parent.

![2018-10-27-013545_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597179-b7f90800-d98c-11e8-8ebe-a2d035c00721.png)

This is where it goes wrong. Really wrong. As the OperandBundles and InstMetadata of the *previous* CallInst are added to the current CallInst.

The only reason this happens is since OperandBundles and InstMetadata are zero in length, thus offset if not enough to distinguish which parent nonterminal they belong to.

![2018-10-27-013740_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597181-bb8c8f00-d98c-11e8-9021-9461ff0e4849.png)

This is the incorrect firstChild. It points to InstMetadata, but should really point to

![2018-10-27-013942_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597184-bf201600-d98c-11e8-9b04-81157c59a48b.png)

Incorrect parents have now been set for InstMetadata and OperandBundles.

![2018-10-27-014217_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597190-c34c3380-d98c-11e8-9ad0-352f8f8d4432.png)

No parent was set for OperandBundles and InstMetadata. They should have both been assigned the last CallInst as parent (i.e. index 18).

![2018-10-27-014315_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597192-c6472400-d98c-11e8-8f97-74f7d5e80776.png)

Since InstMetadata (at index 10) is now the firstChild of the CallInst at index 18, and since the InstMetadata does not have a next (i.e. next is 0), the CallInst will not have any other children than the incorrectly added InstMetadata child. Therefore, calling Typ on CallInst will result in NONE type.

![2018-10-27-014532_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597194-c9421480-d98c-11e8-9b69-7bbdc79d9b5b.png)

InstMetadata (at index 17) of previous instruction being added to have RetTerm as parent. The correct InstMetadata to add to RetTerm is at index 20.

![2018-10-27-014915_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597196-cba46e80-d98c-11e8-8154-c584e05f8f1c.png)

FuncMetadata incorrectly added to have FuncBody as parent.

![2018-10-27-015614_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597197-cfd08c00-d98c-11e8-9130-a49667e1cbb9.png)

InstMetadata incorrectly added to have FuncBody as parent (should have been part of TermRet).

![2018-10-27-015730_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597199-d3641300-d98c-11e8-9c4d-e7c8241525ab.png)

Panic with `unknown node type NONE`, since CallInst has InstMetadata as first child.

![2018-10-27-020137_1920x1175_scrot](https://user-images.githubusercontent.com/1414531/47597204-d65f0380-d98c-11e8-8186-fdb93f7395cd.png)

**Comment by inspirer** [comment]
Yeah, this is working as expected with the given data model. The parser produces a bunch of ranges, and the AST builder looks at how they are nested to produce the correct structure. This breaks profoundly with nullable nonterminals. First of all, the parser cannot properly position a nullable nonterminal in source code, and just puts it right before the next token, consider:
```
  input : A d ; 
  A : b copt ;
```
and `b d` as an input. There are two possibilities for positioning copt: right after `b` and right before `d`. Textmapper always picks the latter, which leads to including trailing spaces in both `copt` and `A`. The proper positioning depends on the outer rules, and we don't know which rule is going to be applied.

There are two solutions to this problem:
- To avoid (forbid?) nullable nonterminals at first and last positions in a rule. This is the rule I'm currently following everywhere. Textmapper should flag when a nullable terminal is used in such a way (a TODO for me).
- Propagate more information into the AST builder so that it can adjust offsets for empty nodes. This is a very valid approach but it comes with performance penalties I'd like to avoid (we can/should still experiment and measure but my gut feeling says that avoiding nullable terminals in the first place is a much cheaper and better approach).

The rewritten grammar should look like this:
```
input : A d ;
A : b c? ;
```

For now, I'll use this issue to work on flagging occurrences of nullable nonterminals in unwanted locations in event based-parsers.

**Comment by mewmew** [comment]
> To avoid (forbid?) nullable nonterminals at first and last positions in a rule. This is the rule I'm currently following everywhere. Textmapper should flag when a nullable terminal is used in such a way (a TODO for me).

I see your point and I was thinking of this too. This is definitely the simplest path to take for TextMapper. It may however put additional constraints on TM grammar writers. Specifically for the grammar of LLVM IR, optional nonterminals as used *everywhere*. So for this grammar, they would have to be rewritten to use `Fooopt` or `Foo?` or `Bar*` or `Bar+?` instead of nullable nonterminals.

[GlobalDecl](https://github.com/mewmew/l-tm/blob/e5db8cb9ff13090ab73125afdea85dabad883d8e/asm/ll/ll.tm#L811) uses this approach:

```c
GlobalDecl -> GlobalDecl
   : Name=GlobalIdent '=' ... GlobalAttrs=(',' GlobalAttr)+? FuncAttrs=(',' FuncAttr)+?
;
```

It would break for the current implementation of [FuncDef](https://github.com/mewmew/l-tm/blob/e5db8cb9ff13090ab73125afdea85dabad883d8e/asm/ll/ll.tm#L889) I believe as `FuncMetadata could belong to either `FuncHeader`, `FuncBody` or `FuncDecl`. With the current implementation it would belong to `FuncBody` right? It should belong to `FuncDef`.

```c
FuncDef -> FuncDef
	: 'define' Header=FuncHeader Metadata=FuncMetadata Body=FuncBody
;
```

The solution being to inline the definition of [FuncMetadata](https://github.com/mewmew/l-tm/blob/e5db8cb9ff13090ab73125afdea85dabad883d8e/asm/ll/ll.tm#L4474) wherever used.

```c
FuncMetadata -> FuncMetadata
	: MetadataAttachments=MetadataAttachment*
;
```

Just to give a sense for how much LLVM IR uses optional nonterminals, below follows the definition of [FuncHeader](https://github.com/mewmew/l-tm/blob/e5db8cb9ff13090ab73125afdea85dabad883d8e/asm/ll/ll.tm#L903):

```c
FuncHeader -> FuncHeader
	: (Linkage | ExternLinkage)? Preemptionopt Visibilityopt DLLStorageClassopt CallingConvopt ReturnAttrs=ReturnAttr* RetType=Type Name=GlobalIdent '(' Params ')' UnnamedAddropt AddrSpaceopt FuncAttrs=FuncAttr* Sectionopt Comdatopt GCopt Prefixopt Prologueopt Personalityopt
;
```

I mainly give these specific usage examples to have a concrete grammar as a reference when deciding on either approach 1) or 2). I too would like to keep things simple. Still, the *intuitive* approach as a user of TextMapper would be if nonterminals belong to their respective parent, as defined in the grammar. And this would be approach 2.

Kindly,
Robin

**Comment by inspirer** [comment]
I think we can introduce a generation option have both, with 2 being the default. With the go rewrite, we will get declarative inlining, which should make a lot of things easier.
```
FuncMetadata inline
	: (MetadataAttachments=MetadataAttachment+  -> FuncMetadata)?
;
```

**Comment by inspirer** [comment]
Yes, inlining is not always for free and I agree that we should have an option to patch locations for nullable nonterminals. I'll work on this.

**Comment by mewmew** [comment]
Just to add to the discussion, if approach 2 tracks the parent of each non-terminal, perhaps such a solution could also allow for non-terminals to appear at different locations within a production rule. Once more, approach 1 would favour simplicity (and speed) whereas approach 2 could favour intuitive use.

Essentially, as a user of Textmapper, I'd be fine with such trade-offs. There may of course be some fundamental aspects of how the shift/reduce parser operates that makes it not possible to do this for approach 2, if so I'd be glad to learn about them.

For a concrete example of the above that hit us today, see the following grammar:

```perl
FuncDecl -> FuncDecl
	: 'declare' Metadata=MetadataAttachment* Header=FuncHeader
	| 'define' Header=FuncHeader Metadata=MetadataAttachment* Body=FuncBody
;
```

Which results in the following Textmapper error:

```
ll.tm,966: named elements must occur in the same order in all productions
```

Cheers,
Robin
